### PR TITLE
Throttler homework

### PR DESCRIPTION
**Pull request by TomaszRewak**
My implementation of throttler

Basic usage:
```
auto throttler = make_message_throttler<UserId, 4, Message>(message_consumer);

throttler.from(1).send(1);
throttler.from(1).send(2);

throttler.from(2).send(3);
```
or
```
auto throttler = make_message_throttler<UserId, 4, Message>(message_consumer);

auto& client_interface = throttler.from(1);
client_interface.send(1)
client_interface.send(2);

throttler.from(2).send(3);
```

More advanced usage:
```
auto throttler = make_message_throttler<UserId, 4, Message>(
	message_consumer, 
	disposed_message_consumer, 
	chrono_timestamp_threshold{ std::chrono::milliseconds{ 1000 } });
```
or 
```
auto throttler = make_message_throttler<UserId, 4, Message, Timestamp>(
	message_consumer, 
	disposed_messages_consumer, 
	custom_sliding_window_size, 
	custom_timestamper);
```

Properties:
- I've separated single client interface from the main throttler. Thanks to this if one client wants to send multiple messages only one map lookup is required.
- Throttler supports custom message types and client id types.
- Complexity of the `send` method (of the client interface) is guaranteed to be O(1) no matter how long the sliding window is and how many messages are already registered in it. Also, after sending N (= sliding window size) messages the list of instructions performed to send [dispose] messages is the always the same (which should make it easy for branch prediction algorithm).
- `send` method (of the client interface) never allocates.
- User can define his own data consumers. He can also listen for discarded messages.
- By default (defined in `message_throttler_commons.hpp`) throttler uses chrono for timestamping. This behavior can be overwritten with any custom timestamper and sliding window processor.
- Calling the throttler's `from` method hashes the clientId only once. Also, map lookup is performed only once, even if the given client has not been registered yet.

### CODE REVIEW

File: course02/homework02/Tomasz/message_throttler.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		const _TimestampThreshold& timestampThreshold) :
+		mMessageConsumer(messageConsumer),
+		mMessageDisposer(messageDisposer),
+		mTimestamper(timestamper),
+		mTimestampThreshold(timestampThreshold)
+	{ }
+
+	_MessageThrottlerInterface& from(const _ClientId& clientId)
+	{
+		auto element = mClients.emplace(clientId, nullptr);
```
Comment (david-grs) [inline course02/homework02/Tomasz/message_throttler.hpp]: Calling this the element is misleading; its the pair<it, bool>, not the element.

File: course02/homework02/Tomasz/message_throttler.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		auto& client = element.first->second;
+
+		if (element.second)
```
Comment (david-grs) [inline course02/homework02/Tomasz/message_throttler.hpp]: Using first/second on multiple lines is not the best, it's hard to read and reason about it.

File: course02/homework02/Tomasz/message_throttler.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			client = std::make_unique<_MessageThrottlerInterface>(mMessageConsumer, mMessageDisposer, mTimestamper, mTimestampThreshold);
```
Comment (david-grs) [inline course02/homework02/Tomasz/message_throttler.hpp]: I find the pattern of inserting a nullptr, and modifying it to a unique_ptr if inserted is a bit complicated. Why not just

```
auto& client = mClients[clientId];
if (client == nullptr)
    client = std::make_unique...
return client;
```

Although I must admit that I prefer an APIwhere clients would be explicitly announced through another function, and her we would always expect to `find` a client without creating it at the same time

File: course02/homework02/Tomasz/short_circular_buffer.hpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <array>
+#include <cassert>
+
+template<typename _Ty, std::size_t _Size>
+class short_circular_buffer
```
Comment (david-grs) [inline course02/homework02/Tomasz/short_circular_buffer.hpp]: Nice, but it already exists, see boost.circular buffer

File: course02/homework02/Tomasz/message_throttler_commons.hpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	std::size_t _BufferSize,
+	typename _Message,
+	typename _Timestamp = std::chrono::milliseconds,
+	typename _MessageConsumer = message_swallower<_Message>,
+	typename _MessageDisposer = message_swallower<_Message>,
+	typename _Timestamper = chrono_timestamper,
+	typename _TimestampThreshold = chrono_timestamp_threshold,
+	typename _Buffer = short_circular_buffer<_Timestamp, _BufferSize>,
+	typename _MessageThrottlerInterface = message_throttler_interface<_Message, _MessageConsumer, _MessageDisposer, _Timestamp, _Timestamper, _TimestampThreshold, _Buffer>,
+	typename _Map = std::map<_ClientId, std::unique_ptr<_MessageThrottlerInterface>>
```
Comment (david-grs) [inline course02/homework02/Tomasz/message_throttler_commons.hpp]: It's nice to template for customization points, although I must admit that you took that a bit on the extreme side :) Typically, the map isn't really needed: If you have one class SlidingWindow<Message, ...> with most of the logic, the outside class should be dead simple - just an unordered_map of client to SlidingWindow.  Which is why it doesnt need to be templated - the client code can just be changed. Timestamp and Timestamper can be replaced if you template on the Clock type - then you know your time point/duration/etc types and the related operations. BufferSize and TimestampThreshold should be parameters, not compile time constants imho.

File: course02/homework02/Tomasz/message_throttler_interface.hpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+template<
+	typename _Message,
+	typename _MessageConsumer,
+	typename _MessageDisposer,
+	typename _Timestamp,
+	typename _Timestamper,
+	typename _TimestampThreshold,
+	typename _Buffer
+>
+class message_throttler_interface
```
Comment (david-grs) [inline course02/homework02/Tomasz/message_throttler_interface.hpp:11]: Why calling it interface?

File: course02/homework02/Tomasz/main.cc [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+	throttler_test();
+	throttler_dispose_test();
+	throttler_single_interface_test();
+	throttler_multiple_users_test();
+	throttler_timeout_test();
+	throttler_custom_timestamper_test();
+	throttler_chrono_test();
+	throttler_chrono_partial_test();
+	throttler_chrono_full_cycle_test();
+	throttler_chrono_muliple_users_test();
```
Comment (david-grs) [inline course02/homework02/Tomasz/main.cc:251]: Very nice!

**Review by david-grs — APPROVED** [review]
Nice implementation!

File: course02/homework02/Tomasz/message_throttler.hpp (hunk 3 above)
Comment (TomaszRewak) [reply course02/homework02/Tomasz/message_throttler.hpp]: Yeah, the history of this line is quite old :P 
You are completely, this pattern is much cleaner.

File: course02/homework02/Tomasz/short_circular_buffer.hpp (hunk 4 above)
Comment (TomaszRewak) [reply course02/homework02/Tomasz/short_circular_buffer.hpp]: Next time will definitely use that one! Thanks.

Comment (TomaszRewak) [reply course02/homework02/Tomasz/short_circular_buffer.hpp]: Or actually it's probably time to learn boost. I will change that.

File: course02/homework02/Tomasz/message_throttler_commons.hpp (hunk 5 above)
Comment (TomaszRewak) [reply course02/homework02/Tomasz/message_throttler_commons.hpp]: I've changed all of the _Timestampe[Threshold|er] parameters with just a single _Clock parameter. And I also removed _Map from parameters.

File: course02/homework02/Tomasz/message_throttler_interface.hpp (hunk 6 above)
Comment (TomaszRewak) [reply course02/homework02/Tomasz/message_throttler_interface.hpp:11]: It was the best name that came to my mind as it gives access to the throttler for a single user.

**Pull request merged**